File: crates/core/benches/hierarchy.rs
//! Hierarchy-mode (`->`) benchmarks (`hl7pet_core::execute_hierarchy`) — the
//! Rust equivalent of spec 004's new `HierarchyBenchmarks.java` (spec 009
//! research.md #4), run against the same corpus entry that Scala side
//! benchmarks.
//!
//! Deliberately scoped to exactly the corpus's hierarchy-eligible entries
//! (currently just `large_hierarchy_028`) — NOT also the existing
//! `basic-hierarchy.hl7`/`complex-hierarchy.hl7` correctness fixtures an
//! earlier task draft considered reusing here. Scala's `HierarchyBenchmarks`
//! only benchmarks `large-hierarchy.hl7`; adding extra Rust-only messages
//! would produce Comparison Result rows with no Scala counterpart, breaking
//! FR-001's "exact same named corpus messages" guarantee for no real
//! benefit (those two fixtures are tiny — 9 and 12 lines — and add no scale
//! signal `large-hierarchy.hl7` doesn't already provide).

#[path = "common/mod.rs"]
mod common;

use common::output::{write_partial_results, MetricValue, ResultRow};
use common::timing::{measure_operation, DEFAULT_MEASURED_ITERS, DEFAULT_WARMUP_ITERS};
use hl7pet_core::HierarchyProfile;

/// Mirrors `HierarchyBenchmarks.java`'s four representative PATH forms
/// exactly, for a true row-for-row comparison.
const PATH_FORMS: &[&str] =
    &["OBR[1] -> OBX-5", "OBR[1] -> OBX[3]-5", "OBR[1] -> OBX[@5='VAL-1-2']-5", "OBR -> OBX-5"];

fn main() {
    let corpus = common::corpus::load();
    let mut results = Vec::new();
    let mut engine_failures = Vec::new();

    for message in corpus.hierarchy_eligible() {
        let profile_json = message.profile_json.as_deref().expect("hierarchy_eligible guarantees Some");

        let scan_result = match hl7pet_core::scan(&message.content) {
            Ok(s) => s,
            Err(e) => {
                engine_failures.push(common::output::EngineFailure {
                    engine: "rust",
                    message_id: message.message_id.clone(),
                    feature: "hierarchy",
                    description: format!("scan() failed: {e}"),
                });
                continue;
            }
        };
        let profile = match HierarchyProfile::from_json(profile_json) {
            Ok(p) => p,
            Err(e) => {
                engine_failures.push(common::output::EngineFailure {
                    engine: "rust",
                    message_id: message.message_id.clone(),
                    feature: "hierarchy",
                    description: format!("HierarchyProfile::from_json() failed: {e}"),
                });
                continue;
            }
        };

        for path_expr in PATH_FORMS {
            let compiled = match hl7pet_core::parse(path_expr) {
                Ok(c) => c,
                Err(e) => {
                    engine_failures.push(common::output::EngineFailure {
                        engine: "rust",
                        message_id: message.message_id.clone(),
                        feature: "hierarchy",
                        description: format!("parse({path_expr:?}) failed: {e}"),
                    });
                    continue;
                }
            };

            let (timing, alloc) = measure_operation(DEFAULT_WARMUP_ITERS, DEFAULT_MEASURED_ITERS, || {
                hl7pet_core::execute_hierarchy(&scan_result, &compiled, Some(&profile))
            });

            results.push(ResultRow {
                feature: "hierarchy",
                message_id: message.message_id.clone(),
                path_expression: Some(path_expr.to_string()),
                throughput: MetricValue { value: timing.throughput_ops_per_us, unit: "ops/us" },
                latency_p50: MetricValue { value: timing.p50_us, unit: "us" },
                latency_p95: MetricValue { value: timing.p95_us, unit: "us" },
                allocation_bytes_per_op: MetricValue { value: alloc.bytes_per_op, unit: "bytes" },
                memory_alloc_rate_bytes_per_sec: MetricValue {
                    value: alloc.bytes_per_op * timing.throughput_ops_per_us * 1_000_000.0,
                    unit: "bytes/sec",
                },
                allocation_call_count: MetricValue { value: alloc.call_count_per_op, unit: "count" },
            });
        }
    }

    write_partial_results("hierarchy", corpus.corpus_id, results, engine_failures);
}

File: crates/python/benches/hierarchy.py
"""`get_value_hierarchy` benchmarks (spec 6001-python-ffi-benchmark) --
reuses `crates/core/benches/hierarchy.rs`'s exact four-form `PATH_FORMS`
constant against `large_hierarchy_028`, calling the `hl7pet` Python
binding instead of `hl7pet_core` directly.
"""

from __future__ import annotations

import json

import hl7pet

from common.corpus import Corpus
from common.output import EngineFailure, MetricValue, ResultRow
from common.timing import sample

# Mirrors hierarchy.rs's PATH_FORMS exactly, for a true row-for-row
# comparison.
PATH_FORMS = (
    "OBR[1] -> OBX-5",
    "OBR[1] -> OBX[3]-5",
    "OBR[1] -> OBX[@5='VAL-1-2']-5",
    "OBR -> OBX-5",
)


def run(corpus: Corpus) -> tuple[list[ResultRow], list[EngineFailure]]:
    results: list[ResultRow] = []
    failures: list[EngineFailure] = []

    for message in corpus.hierarchy_eligible():
        assert message.profile_json is not None  # hierarchy_eligible() guarantees this
        try:
            profile = json.loads(message.profile_json)
        except json.JSONDecodeError as e:
            failures.append(
                EngineFailure(
                    message_id=message.message_id,
                    feature="hierarchy",
                    description=f"profile JSON failed to parse: {e}",
                )
            )
            continue

        for path_expr in PATH_FORMS:
            try:
                stats = sample(lambda: hl7pet.get_value_hierarchy(message.content, path_expr, profile))
            except Exception as e:  # noqa: BLE001
                failures.append(
                    EngineFailure(
                        message_id=message.message_id,
                        feature="hierarchy",
                        description=f"get_value_hierarchy(message, {path_expr!r}, profile) failed: {e}",
                    )
                )
                continue

            results.append(
                ResultRow(
                    feature="hierarchy",
                    message_id=message.message_id,
                    path_expression=path_expr,
                    throughput=MetricValue(stats.throughput_ops_per_us, "ops/us"),
                    latency_p50=MetricValue(stats.p50_us, "us"),
                    latency_p95=MetricValue(stats.p95_us, "us"),
                )
            )

    return results, failures
